<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [370]:
import pandas as pd
import psycopg2

In [371]:
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [372]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [373]:
# текст запроса
query_3_1 = f'''
select
    count(*)
from vacancies
'''

In [374]:
df_3_1=pd.read_sql_query(query_3_1, connection)
df_3_1

C:\Users\Анастасия\AppData\Local\Temp\ipykernel_15980\4190961098.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_1=pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [375]:
# текст запроса
query_3_2 = f'''
select
    count(*)
from employers
'''

In [376]:
df_3_2=pd.read_sql_query(query_3_2, connection)
df_3_2

C:\Users\Анастасия\AppData\Local\Temp\ipykernel_15980\2139395580.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_2=pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [377]:
# текст запроса
query_3_3= f'''
select
    count(*)
from areas
'''

In [378]:
df_3_3=pd.read_sql(query_3_3, connection)
df_3_3

C:\Users\Анастасия\AppData\Local\Temp\ipykernel_15980\2806143467.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_3=pd.read_sql(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [379]:
# текст запроса
query_3_4 = f'''
select
    count(*)
from industries
'''

In [380]:
df_3_4=pd.read_sql(query_3_4, connection)
df_3_4

C:\Users\Анастасия\AppData\Local\Temp\ipykernel_15980\1218951428.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_4=pd.read_sql(query_3_4, connection)


count
0    294

По итогу имеем: 49197 вакансий, 23501 работадателя в 1362 редионах и 294 сферы деятельности

***

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [381]:
query_4_1=f'''
select  
        a.name, 
        count(v.id) --количество вакансий
from VACANCIES v
join AREAS a on v.area_id=a.id -- объединяем таблици по ключу
group by a.name              --группируем по названию региона
order by count(v.id) desc     -- сортируем по количеству вакансий
limit 10
'''



In [382]:
df_4_1=pd.read_sql(query_4_1, connection)
df_4_1

C:\Users\Анастасия\AppData\Local\Temp\ipykernel_15980\2597679622.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_1=pd.read_sql(query_4_1, connection)


name  count
0           Москва   5333
1  Санкт-Петербург   2851
2            Минск   2112
3      Новосибирск   2006
4           Алматы   1892
5     Екатеринбург   1698
6  Нижний Новгород   1670
7           Казань   1415
8        Краснодар   1301
9           Самара   1144

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [383]:
query_4_2='''
select
        count(id)
from VACANCIES
where (salary_to is not null) or (salary_from is not null) -- прописываем условие, существования хотябы одного из двух полей
'''

In [384]:
df_4_2=pd.read_sql(query_4_2, connection)
df_4_2

C:\Users\Анастасия\AppData\Local\Temp\ipykernel_15980\3928226794.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_2=pd.read_sql(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [385]:
query_4_3='''
select
       round( avg(salary_to)) l,
        round(avg(salary_from)) u
from VACANCIES
where (salary_to is not null) or (salary_from is not null) -- прописываем условие, существования хотябы одного из двух полей


'''

In [386]:
df_4_3=pd.read_sql(query_4_3, connection)
df_4_3

C:\Users\Анастасия\AppData\Local\Temp\ipykernel_15980\1912213825.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_3=pd.read_sql(query_4_3, connection)


l        u
0  110537.0  71065.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [387]:
query_4_4='''
select
        schedule,
        employment,
        count(id)
from VACANCIES
group by schedule,employment --группируем по типу графика и типу трудоустройства
order by count(id) desc -- сортируем
'''

In [388]:
df_4_4=pd.read_sql(query_4_4, connection)
df_4_4

C:\Users\Анастасия\AppData\Local\Temp\ipykernel_15980\2066619207.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_4=pd.read_sql(query_4_4, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [389]:
query_4_5='''
select
        experience
from VACANCIES
group by experience --группируем по опыту
order by count(id) --сортируем в порядке возрастания кол-ва вокансий
'''

In [390]:
df_4_5=pd.read_sql(query_4_5, connection)
df_4_5

C:\Users\Анастасия\AppData\Local\Temp\ipykernel_15980\2595987021.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_5=pd.read_sql(query_4_5, connection)


experience
0         Более 6 лет
1           Нет опыта
2       От 3 до 6 лет
3  От 1 года до 3 лет

***

Нам удалось узнать, что самое большое количество вакансий в больших городах: Москва-5333, Санкт-Петербург-2851, Минск-2112. У 24073 кол-ва вакансий заполненщ хотя бы одно из двух полей с зарплатой. У вурхней граници заработной платы среднее -110537. а у нижней 71065. Количество вакансий для каждого сочетания типа рабочего графика и трудоустройства самым популярным стало: Полный день-Полная занятость. Самым популярным требуемым опытом работы стал опыт: от 1 года до 3 лет.


# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [391]:
query_5_1='''
(select
        e.name,
        count(v.id)
from VACANCIES v
join EMPLOYERS e on e.id=v.employer_id -- соединяем таблицы по ключу
group by e.name                     -- группируем по ключу
order by count(v.id) desc            -- сортируем по кол-ву вакансий 
limit 1)                    --выводим только первый элемент
union                      --дополняем 
(select
        e.name,
        count(v.id)
from VACANCIES v
join EMPLOYERS e on e.id=v.employer_id
group by e.name
order by count(v.id) desc
offset 4 limit 1)             -- выводим 5 элемент
order by 2 desc


'''

In [392]:
df_5_1=pd.read_sql_query(query_5_1, connection)
df_5_1

C:\Users\Анастасия\AppData\Local\Temp\ipykernel_15980\970819196.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_1=pd.read_sql_query(query_5_1, connection)


name  count
0         Яндекс   1933
1  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [393]:
query_5_2=''' 
select
        a.name,
        count(e.id) employer,
        count(v.id) vacancies
from areas a
left join VACANCIES v on a.id=v.area_id -- присоединяем таблицу
left join EMPLOYERS e on a.id=e.area   -- присоединяем таблицу
where v.id is null                       -- выбираем ячейки с нулём
group by a.id                       -- группируем по id
order by employer desc       --сортируем по убыванию работодателей
'''

In [394]:
df_5_2=pd.read_sql_query(query_5_2, connection)
df_5_2

C:\Users\Анастасия\AppData\Local\Temp\ipykernel_15980\2929453857.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_2=pd.read_sql_query(query_5_2, connection)


name  employer  vacancies
0                Россия       410          0
1             Казахстан       207          0
2    Московская область        75          0
3    Краснодарский край        19          0
4    Ростовская область        18          0
..                  ...       ...        ...
588             Скидель         0          0
589          Белоозерск         0          0
590           Орловский         0          0
591          Павловская         0          0
592             Таксимо         0          0

[593 rows x 3 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [395]:
query_5_3=''' 
select
        e.name,
        count(distinct v.area_id) --количество униикальных областей
from VACANCIES v 
join EMPLOYERS e on e.id=v.employer_id -- соединяем две таблицы по клчам
group by e.name       -- групируем по имени работодателя
order by count( v.area_id) desc--сортируем
'''

In [396]:
df_5_3=pd.read_sql_query(query_5_3, connection)
df_5_3

C:\Users\Анастасия\AppData\Local\Temp\ipykernel_15980\756883519.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_3=pd.read_sql_query(query_5_3, connection)


name  count
0                                     Яндекс    181
1                                 Ростелеком    152
2                                   Тинькофф     43
3                                       СБЕР     24
4                              Газпром нефть     24
...                                      ...    ...
14761             ПСБ Инновации и Инвестиции      1
14762                             ПСЗ Янтарь      1
14763                                    ПСК      1
14764  Псковский завод механических приводов      1
14765                     ПСК ТатЭнергоСтрой      1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [397]:
query_5_4=''' 
select 
        count(e.id)
from employers e
left join employers_industries ei on ei.employer_id=e.id -- присоединяем таблицу по ключу
where industry_id is NULL       --выбираем ячейки с нулём

'''

In [398]:
df_5_4=pd.read_sql_query(query_5_4, connection)
df_5_4

C:\Users\Анастасия\AppData\Local\Temp\ipykernel_15980\1578107129.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_4=pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [399]:
query_5_5=''' 
select 
        e.name,
        count(e.id)
from employers e
left join employers_industries ei on ei.employer_id=e.id -- присоединяем таблицу
group by e.name                                           -- группируем по имени
having count(e.id)=4                       -- выбираем компании у которых указано 4 сферы деятельности
order by e.name                            -- сортируем по названию компании
offset 2 limit 1 
'''

In [400]:
df_5_5=pd.read_sql_query(query_5_5, connection)
df_5_5

C:\Users\Анастасия\AppData\Local\Temp\ipykernel_15980\1616719299.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_5=pd.read_sql_query(query_5_5, connection)


name  count
0  2ГИС      4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [401]:
query_5_6=''' 
select
        count(ei.employer_id)
from employers_industries ei 
join industries i on i.id=ei.industry_id        -- соединяем таблицы по ключу
where i.name like 'Разработка программного обеспечения' -- выбираем нужную сферу деятельности

'''

In [402]:
df_5_6=pd.read_sql_query(query_5_6, connection)
df_5_6

C:\Users\Анастасия\AppData\Local\Temp\ipykernel_15980\2046112031.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_6=pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [404]:
# код для получения списка городов-милионников
from urllib.parse import quote_plus
url = quote_plus('https://ru.wikipedia.org/wiki/Города-миллионеры_России', safe=':/')
tables = pd.read_html(url)
mln_cities = tables[1]['Город'].values.tolist()

In [407]:

query_5_7 = f'''
-- Выбираем данные о регионах и количестве вакансий
select
a.name as region,
count(*) as vacancies_count -- считаем количество вакансий
from public.vacancies as v
join public.employers as e on v.employer_id = e.id
join public.areas as a on v.area_id = a.id
group by v.area_id, a.name, e.id, e.name -- группируем по area_id, названию региона, employer_id и названию работодателя
having a.name in {tuple(mln_cities)} and e.name = 'Яндекс' -- применяем фильтры по городам из множества mln_cities и работодателю Яндекс

union all

-- Вторая часть запроса для получения общего количества вакансий

select
'Total', -- вместо региона подставляем строку 'Total'
sum(x.cnt) -- суммируем количество вакансий из подзапроса x
from
(
select
count(*) as cnt -- подзапрос для подсчета количества вакансий
from
public.vacancies as v
join public.employers as e on v.employer_id = e.id
join public.areas as a on v.area_id = a.id
group by v.area_id, a.name, e.id, e.name -- группируем по тем же столбцам
having a.name in {tuple(mln_cities)} and e.name = 'Яндекс' -- применяем те же фильтры
) as x

order by 2 -- Сортируем результат по второму столбцу (количеству вакансий)
'''


In [408]:
df_5_7=pd.read_sql_query(query_5_7, connection)
df_5_7

C:\Users\Анастасия\AppData\Local\Temp\ipykernel_15980\2818792313.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_5_7=pd.read_sql_query(query_5_7, connection)


region  vacancies_count
0              Омск             21.0
1         Челябинск             22.0
2        Красноярск             23.0
3         Волгоград             24.0
4    Ростов-на-Дону             25.0
5            Казань             25.0
6             Пермь             25.0
7               Уфа             26.0
8            Самара             26.0
9         Краснодар             30.0
10          Воронеж             32.0
11      Новосибирск             35.0
12  Нижний Новгород             36.0
13     Екатеринбург             39.0
14  Санкт-Петербург             42.0
15           Москва             54.0
16            Total            485.0

In [ ]:
# результат запроса

Из полученных данных понятно, что лидирами среди работодателей являются Газпром и Яндекс. Яндекс распространяется на 181 регион, и 485 вакансий
. Многие работодатели не указывают сыеру деятельности что припятствует анализу. 

***

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [ ]:
query_6_1=''' 
select 
        count(name)
from vacancies
where lower(name) like '%data%' or lower(name) like '%данн%' -- выбираем данные содержащие нужные нам слова
''' 

In [ ]:
df_6_1=pd.read_sql_query(query_6_1, connection)
df_6_1

C:\Users\Анастасия\AppData\Local\Temp\ipykernel_15980\3821253376.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_1=pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [ ]:
query_6_2=''' 
select
        count(*)
from vacancies
where (lower(name) like '%data scientist%'  -- выбираем данные содержащие нужные нам слова
or lower(name) like '%data science%' 
or lower(name) like '%исследователь данных%' 
or lower(name) like '%machine learning%' 
or lower(name) like '%машинн%обучен%' 
or (name like '%ML%' and lower(name) not like '%html%'))  
and
(lower(name) like '%junior%' 
or lower(experience) like '%нет опыта%' 
or lower(employment) like '%cтажировка%')

        
        
'''

In [ ]:
df_6_2=pd.read_sql_query(query_6_2, connection)
df_6_2

C:\Users\Анастасия\AppData\Local\Temp\ipykernel_15980\1590952481.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_2=pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [ ]:
query_6_3=''' 
select
        count(*)
from vacancies
where (lower(name) like '%data scientist%'  -- выбираем данные содержащие нужные нам слова
or lower(name) like '%data science%' 
or lower(name) like '%исследователь данных%' 
or lower(name) like '%machine learning%' 
or lower(name) like '%машинн%обучен%' 
or (name like '%ML%' and lower(name) not like '%html%'))
and
(
    lower(key_skills) like '%sql%'
    or lower(key_skills) like '%postgres%'
)

        
        
'''

In [ ]:
df_6_3=pd.read_sql_query(query_6_3
                         , connection)
df_6_3

C:\Users\Анастасия\AppData\Local\Temp\ipykernel_15980\1425101736.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_3=pd.read_sql_query(query_6_3


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [ ]:
query_6_4=''' 
select
        count(*)
from vacancies
where (lower(name) like '%data scientist%'   -- выбираем данные содержащие нужные нам слова
or lower(name) like '%data science%' 
or lower(name) like '%исследователь данных%' 
or lower(name) like '%machine learning%' 
or lower(name) like '%машинн%обучен%' 
or (name like '%ML%' and lower(name) not like '%html%'))
and
(
    lower(key_skills) like '%python%'
)
'''

In [ ]:
df_6_4=pd.read_sql_query(query_6_4
                         , connection)
df_6_4

C:\Users\Анастасия\AppData\Local\Temp\ipykernel_15980\1971380948.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_4=pd.read_sql_query(query_6_4


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [ ]:
query_6_5 = f'''
select
        round(avg(length(key_skills) - length(replace(key_skills, '\t', '')) + 1), 2) --средняя длина списка навыков в символах , учитывая удаление символов табуляции, и округляет результат до двух знаков после запятой.
from VACANCIES
where (lower(name) like '%data scientist%' or -- выбираем данные содержащие нужные нам слова
lower(name) like '%data science%' or
lower(name) like '%исследователь данных%' or
lower(name) like '%machine learning%' or
lower(name) like '%машинн%обучен%' or
(name like '%ML%' and lower(name) not like '%html%'))
'''

In [ ]:
df_6_5=pd.read_sql_query(query_6_5
                         , connection)
df_6_5

C:\Users\Анастасия\AppData\Local\Temp\ipykernel_15980\4057770753.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_5=pd.read_sql_query(query_6_5


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [ ]:

query_6_6 = f'''
select
        experience,
        round(avg((coalesce(salary_from, salary_to) + coalesce(salary_to, salary_from)) / 2)) avg_sal --средняя зарплатф на основе минимальной и максимальной зарплаты вакансий, при этом учитывая возможное отсутствие данных по одной из границ, и округляет результат до целого числа.

from VACANCIES
where (lower(name) like '%data scientist%' or -- выбираем данные содержащие нужные нам слова
        lower(name) like '%data science%' or
        lower(name) like '%исследователь данных%' or
        lower(name) like '%machine learning%' or
        lower(name) like '%машинн%обучен%' or
(name like '%ML%' and lower(name) not like '%html%'))
and
(salary_from is not null or salary_to is not null) --условие наличия данных хотя бы для минимальной или максимальнной зарплаты
GROUP BY experience -- группируем по опыту
ORDER BY avg_sal DESC
'''

In [ ]:
df_6_6=pd.read_sql_query(query_6_6
                         , connection)
df_6_6

C:\Users\Анастасия\AppData\Local\Temp\ipykernel_15980\1368942366.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_6_6=pd.read_sql_query(query_6_6


experience   avg_sal
0       От 3 до 6 лет  243115.0
1  От 1 года до 3 лет  139675.0
2           Нет опыта   74643.0

***

Из проанализированных данных следует, что из всех вакансий, только 1771 связаны с данными, и тлько 51 из них - это новичёк дс. Также стало заметно. что владение Питоном и SQL увеличивает ваши шансы на трудоустройство. Также для дс практически обзательно знание питона. Опыт тоже играет большую роль, особенно в формировании зарплаты. Однако только малая часть работодателей указывает размер зарпавты. 

# Общий вывод по проекту

По итогу имеем 1771 вакансию 3,6% из которых, вакансии дата-сайнтиста. Для увеличения трудоустройства следует освоить питон и SQL. Также большую роль играет опыт работы, чем он выше, тем болше вакансий для вас открыты. Также опыт влияет на зароботную платуБ однако многие работадатели её не указывают. Также мы показали многообразие работадателей на рынке- 225.   